### PR TITLE
Add Pod download caching

### LINKED ISSUE #2863: Cache downloads while linting a podspec
I know we recently removed the caching, but this has **substantially** slowed down `pod spec lint / pod trunk push` when a podspec has subspecs. For every subspec it will download the source again and again along with any dependencies again and again.

Here's an example:

```bash
$ pod spec lint ~/.cocoapods/repos/master/Specs/KFData/1.0.5/KFData.podspec.json --verbose | grep "git clone"
```

The output contains the following:
```
     $ /usr/local/bin/git clone https://github.com/kylef/KFData.git /private/tmp/CocoaPods/Lint/Pods/KFData --single-branch --depth 1 --branch 1.0.5
     $ /usr/local/bin/git clone https://github.com/QueryKit/QueryKit.git /private/tmp/CocoaPods/Lint/Pods/QueryKit --single-branch --depth 1 --branch 0.8.3
     $ /usr/local/bin/git clone https://github.com/kylef/KFData.git /private/tmp/CocoaPods/Lint/Pods/KFData --single-branch --depth 1 --branch 1.0.5
     $ /usr/local/bin/git clone https://github.com/QueryKit/QueryKit.git /private/tmp/CocoaPods/Lint/Pods/QueryKit --single-branch --depth 1 --branch 0.8.3
     $ /usr/local/bin/git clone https://github.com/kylef/KFData.git /private/tmp/CocoaPods/Lint/Pods/KFData --single-branch --depth 1 --branch 1.0.5
     $ /usr/local/bin/git clone https://github.com/QueryKit/QueryKit.git /private/tmp/CocoaPods/Lint/Pods/QueryKit --single-branch --depth 1 --branch 0.8.3
     $ /usr/local/bin/git clone https://github.com/kylef/KFData.git /private/tmp/CocoaPods/Lint/Pods/KFData --single-branch --depth 1 --branch 1.0.5
     $ /usr/local/bin/git clone https://github.com/QueryKit/QueryKit.git /private/tmp/CocoaPods/Lint/Pods/QueryKit --single-branch --depth 1 --branch 0.8.3
     $ /usr/local/bin/git clone https://github.com/kylef/KFData.git /private/tmp/CocoaPods/Lint/Pods/KFData --single-branch --depth 1 --branch 1.0.5
     $ /usr/local/bin/git clone https://github.com/QueryKit/QueryKit.git /private/tmp/CocoaPods/Lint/Pods/QueryKit --single-branch --depth 1 --branch 0.8.3
     $ /usr/local/bin/git clone https://github.com/kylef/KFData.git /private/tmp/CocoaPods/Lint/Pods/KFData --single-branch --depth 1 --branch 1.0.5
     $ /usr/local/bin/git clone https://github.com/QueryKit/QueryKit.git /private/tmp/CocoaPods/Lint/Pods/QueryKit --single-branch --depth 1 --branch 0.8.3
     $ /usr/local/bin/git clone https://github.com/kylef/KFData.git /private/tmp/CocoaPods/Lint/Pods/KFData --single-branch --depth 1 --branch 1.0.5
     $ /usr/local/bin/git clone https://github.com/kylef/KFData.git /private/tmp/CocoaPods/Lint/Pods/KFData --single-branch --depth 1 --branch 1.0.5
     $ /usr/local/bin/git clone https://github.com/kylef/KFData.git /private/tmp/CocoaPods/Lint/Pods/KFData --single-branch --depth 1 --branch 1.0.5
     $ /usr/local/bin/git clone https://github.com/kylef/KFData.git /private/tmp/CocoaPods/Lint/Pods/KFData --single-branch --depth 1 --branch 1.0.5
     $ /usr/local/bin/git clone https://github.com/kylef/KFData.git /private/tmp/CocoaPods/Lint/Pods/KFData --single-branch --depth 1 --branch 1.0.5
     $ /usr/local/bin/git clone https://github.com/QueryKit/QueryKit.git /private/tmp/CocoaPods/Lint/Pods/QueryKit --single-branch --depth 1 --branch 0.8.3
     $ /usr/local/bin/git clone https://github.com/kylef/KFData.git /private/tmp/CocoaPods/Lint/Pods/KFData --single-branch --depth 1 --branch 1.0.5
     $ /usr/local/bin/git clone https://github.com/QueryKit/QueryKit.git /private/tmp/CocoaPods/Lint/Pods/QueryKit --single-branch --depth 1 --branch 0.8.3
     $ /usr/local/bin/git clone https://github.com/kylef/KFData.git /private/tmp/CocoaPods/Lint/Pods/KFData --single-branch --depth 1 --branch 1.0.5
     $ /usr/local/bin/git clone https://github.com/QueryKit/QueryKit.git /private/tmp/CocoaPods/Lint/Pods/QueryKit --single-branch --depth 1 --branch 0.8.3
     $ /usr/local/bin/git clone https://github.com/kylef/KFData.git /private/tmp/CocoaPods/Lint/Pods/KFData --single-branch --depth 1 --branch 1.0.5
     $ /usr/local/bin/git clone https://github.com/QueryKit/QueryKit.git /private/tmp/CocoaPods/Lint/Pods/QueryKit --single-branch --depth 1 --branch 0.8.3
     $ /usr/local/bin/git clone https://github.com/kylef/KFData.git /private/tmp/CocoaPods/Lint/Pods/KFData --single-branch --depth 1 --branch 1.0.5
     $ /usr/local/bin/git clone https://github.com/QueryKit/QueryKit.git /private/tmp/CocoaPods/Lint/Pods/QueryKit --single-branch --depth 1 --branch 0.8.3
```

That's 26 clones, 15 clones of KFData and 11 clones of QueryKit. If this was cached it would drop this down to 2 clones and substantially speeding up both spec linting and trunk push.

### LINKED ISSUE #3172: Cache non-local Pods
About a year ago, we killed the git cache in CocoaPods for the sake of simplicity, but it's made `pod install` (and the other commands that use it) much slower in certain circumstance.

We ought to create a new cache (like the one in Bundler) that lives inside CocoaPods itself (rather than the downloader), and will cache released pod versions, as well as git commits/tags.

### PR DESCRIPTION
**Pull request by segiddins**
Closes #2863.
Closes #3172.

~~This is still a work-in-progress, but I'm fairly confident it works. The cache itself needs specs, and likely a good refactor. And also docs. But it's a start.~~

- [x] Refactor
- [x] Documentation
- [x] Tests
- [ ] CHANGELOG

\c @orta @kylef @alloy

### CODE REVIEW

File: lib/cocoapods/downloader/cache.rb [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+              if name == spec.name
+                result.location = destination
+                result.spec = spec
+              end
+            end
+          end
+        end
+
+        result
+      rescue Object
```
Comment (kylef) [inline lib/cocoapods/downloader/cache.rb]: Should anyone be raising arbitrary objects instead of `Exception` subclasses?

**Comment by orta** [comment]
As I'm cc'd - we could mention that pod x is getting pulled from cache on a verbose install?

File: lib/cocoapods/downloader/cache.rb (hunk 1 above)
Comment (segiddins) [inline lib/cocoapods/downloader/cache.rb]: People do :)

**Comment by orta** [comment]
@dbgrandi - can you go over this please?

File: lib/cocoapods/config.rb [hunk 2]
```diff
@@ -20,7 +20,7 @@ class Config
       :integrate_targets   => true,
       :new_version_message => true,
 
-      :cache_root          => Pathname.new(Dir.home) + 'Library/Caches/CocoaPods',
+      :cache_root          => Pathname(Dir.home) + 'Library/Caches/CocoaPods',
```
Comment (kylef) [inline lib/cocoapods/config.rb]: I vote we leave this as it was for consistency across CocoaPods

File: lib/cocoapods/sandbox/pod_dir_cleaner.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+      private
+
+      # @return [Array<Sandbox::FileAccessor>] the file accessors for all the
+      #         specifications on their respective platform.
+      #
+      def file_accessors
+        return @file_accessors if @file_accessors
+        @file_accessors = []
+        specs_by_platform.each do |platform, specs|
```
Comment (kylef) [inline lib/cocoapods/sandbox/pod_dir_cleaner.rb]: flat map?

Comment (segiddins) [inline lib/cocoapods/sandbox/pod_dir_cleaner.rb]: This code was copy-pasted from `PodSourceInstaller`

**Comment by segiddins** [comment]
This has now been satisfactorily refactored to be documented + tested.

**Comment by orta** [comment]
niccceeee

**Comment by segiddins** [comment]
Idea: the `slug` in `Pod::Downloader::Request` for released pods should include the hash of the spec, so the cache gets properly invalidated.

**Comment by kylef** [comment]
Remember a spec could reference source via `:branch` etc, so the hash of the spec doesn't always change when the underlying source changes.

**Comment by segiddins** [comment]
@kylef of course, but released specs always point to the same thing -- they're not allowed to point to bare branches, for instance.

File: lib/cocoapods/downloader/cache.rb [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        @root = root
+        root.mkpath unless root.exist?
+      end
+
+      def download_pod(request)
+        cached_pod(request) || uncached_pod(request)
+      rescue Informative
+        raise
+      rescue
+        UI.notice("Error installing #{request.name}")
```
Comment (kylef) [inline lib/cocoapods/downloader/cache.rb]: Maybe we can provide a bit more information in this error (perhaps whatever the exception is?).

As a user if I see `Error installing RestKit` I'd be lost and have no idea where to being making it work again. Would be nice to at least have some kind of hint.

Comment (segiddins) [inline lib/cocoapods/downloader/cache.rb]: @kylef it then re-raises the exception on the next line

File: lib/cocoapods/downloader/cache.rb [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+          result
+        end
+      end
+
+      def download(name, target, params, head)
+        downloader = Downloader.for_target(target, params)
+        if head
+          unless downloader.head_supported?
+            raise Informative, "The pod '" + name + "' does not " \
```
Comment (kylef) [inline lib/cocoapods/downloader/cache.rb]: Any reason for not using:

```diff
- "The pod '" + name + "' does not "
+ "The pod '#{name}' does not "
```

Comment (segiddins) [inline lib/cocoapods/downloader/cache.rb]: copy-paste? I'll change it

File: lib/cocoapods/downloader/cache.rb (hunk 4 above)
Comment (kylef) [inline lib/cocoapods/downloader/cache.rb]: :white_check_mark:

**Comment by 0xmax** [comment]
@segiddins what's the status on this? Would be really nice to have caching again. Is there anything you might need help with?

**Comment by kylef** [comment]
@0xced This feature is planned for the next minor release of CocoaPods. It would be really appreciated if you can pull down this branch and test pod caching in your real world projects.

**Comment by 0xced** [comment]
I guess you meant @0xmax, right?

**Comment by kylef** [comment]
Oops sorry @0xced, for some reason auto-complete preferred you in this context.

**Comment by segiddins** [comment]
This now has slightly more test coverage. Please review?

File: lib/cocoapods/downloader.rb [hunk 6]
```diff
@@ -1,8 +1,48 @@
...
+    #         then the pod will only be cached.
+    #
+    # @param  [Pathname,Nil] cache_path
+    #         the path used to cache pod downloads. If `nil`, then no caching
+    #         will be done.
+    #
+    def self.download(
+      request,
+      target,
+      cache_path: !Config.instance.skip_download_cache && Config.instance.cache_root + 'Pods'
```
Comment (kylef) [inline lib/cocoapods/downloader.rb]: Whoa, this named parameters is pretty cool.

File: CHANGELOG.md [hunk 7]
```diff
@@ -22,6 +22,12 @@ To install release candidates run `[sudo] gem install cocoapods --pre`
   [Samuel Giddins](https://github.com/segiddins)
   [#1241](https://github.com/CocoaPods/CocoaPods/issues/1241)
 
+* Add download caching for pods, which speeds up `pod install` and linting,
+  potentially by several oders of magnitude.
```
Comment (kylef) [inline CHANGELOG.md]: orders*

**Pull request merged**